### DISCUSSION TITLE
Npm test never stops executing

### DISCUSSION THREAD

**Discussion by MauricioRobayo** [community/community#25250] [Actions]
<p>I used <code>create-react-app</code> to bootstrap a new project and left it as it is by default.</p>
<p>Added GitHub actions config for node.js:</p>
<pre><code>name: Build and Deploy
on:
  push:
    branches:
      - master
jobs:
  build-and-deploy:
    runs-on: ubuntu-latest
    steps:
      - name: Checkout
        uses: actions/checkout@master
      - name: Setup Node
        uses: actions/setup-node@v1
        with:
          node-version: '10.x'
      - run: npm ci
      - run: npm test
</code></pre>
<p>The code never passes the <code>npm test</code> step.</p>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/d/dbe80bf5b82b3dc745d44236ad59cf1c083d33b5.jpeg" data-download-href="/uploads/short-url/vnnISoJy43uVnYkVNQQXKacHH3D.jpg?dl=1" title="Annotation 2019-12-06 180003.jpg" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/2121481/181089854-f6625114-e8c1-4f68-872b-0c43a2eda64d.jpg" alt="Annotation 2019-12-06 180003.jpg" data-base62-sha1="vnnISoJy43uVnYkVNQQXKacHH3D" width="690" height="178" srcset="https://aws1.discourse-cdn.com/github/optimized/2X/d/dbe80bf5b82b3dc745d44236ad59cf1c083d33b5_2_690x178.jpg, https://aws1.discourse-cdn.com/github/optimized/2X/d/dbe80bf5b82b3dc745d44236ad59cf1c083d33b5_2_1035x267.jpg 1.5x, https://aws1.discourse-cdn.com/github/optimized/2X/d/dbe80bf5b82b3dc745d44236ad59cf1c083d33b5_2_1380x356.jpg 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/d/dbe80bf5b82b3dc745d44236ad59cf1c083d33b5_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">Annotation 2019-12-06 180003.jpg</span><span class="informations">1696×439 46.3 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>
<p>What am I doing wrong?</p>

**Comment by BrightRan** [comment, marked as answer]
<p>It seems that when you execute the command <strong>npm test</strong> , you should explicitly set the environment variable <strong>CI</strong> , no matter its value is <strong><em>true</em></strong> or <strong><em>false</em></strong>.</p>
<p>You can try the below 2 methods to set this environment variable:</p>
<ul>
<li>
<p>Set CI directly in the command line, for example on Linux, macOS (Bash).</p>
<p>CI=true npm test</p>
</li>
<li>
<p>Define CI as an environment variable in the workflow.</p>
<p>env:<br>
 CI: true</p>
</li>
</ul>
<p>We can define it on workflow level, job level or step level, so that it can be available for different scopes.</p>

**Comment by dbroadhurst** [comment]
<p>Do you need both? CI=true npm test works but just setting the env var hangs when you add another command after the npm test</p>
<pre><code>      - run: npm test
      - run: echo hangs in the line above even though the env var is set
      env:
          CI: true

      - run: CI=true npm test
      - run: echo works and pipeline continues
      env:
          CI: true
</code></pre>
<p>When just using env and I run -run: env I don’t see the CI var</p>

**Comment by kuldip27792** [comment]
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/8/8b95570be91ba6ec70411a160839f5a92fbd921d.png" data-download-href="/uploads/short-url/jUOkkwQIyiyQKuTUHPvwnE92cDX.png?dl=1" title="Screen Shot 2020-04-17 at 12.30.17 AM.png" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/42588373/181089858-3519eb43-0d74-4c5d-b362-3a14165cc84c.png" alt="Screen Shot 2020-04-17 at 12.30.17 AM.png" data-base62-sha1="jUOkkwQIyiyQKuTUHPvwnE92cDX" width="690" height="382" srcset="https://aws1.discourse-cdn.com/github/optimized/2X/8/8b95570be91ba6ec70411a160839f5a92fbd921d_2_690x382.png, https://aws1.discourse-cdn.com/github/optimized/2X/8/8b95570be91ba6ec70411a160839f5a92fbd921d_2_1035x573.png 1.5x, https://aws1.discourse-cdn.com/github/optimized/2X/8/8b95570be91ba6ec70411a160839f5a92fbd921d_2_1380x764.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/8/8b95570be91ba6ec70411a160839f5a92fbd921d_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">Screen Shot 2020-04-17 at 12.30.17 AM.png</span><span class="informations">1490×827 147 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div> <div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/4/4f48ed0636b9521a7e5f4dd857777bb9b031581a.png" data-download-href="/uploads/short-url/bjnXIhSTxvvd3nrNw6aVHbQo8rg.png?dl=1" title="Screen Shot 2020-04-17 at 12.32.06 AM.png" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/42588373/181089859-483f7b61-4aca-4a7d-a092-fb7d9ceaf3b8.png" alt="Screen Shot 2020-04-17 at 12.32.06 AM.png" data-base62-sha1="bjnXIhSTxvvd3nrNw6aVHbQo8rg" width="415" height="500" srcset="https://aws1.discourse-cdn.com/github/optimized/2X/4/4f48ed0636b9521a7e5f4dd857777bb9b031581a_2_415x500.png, https://aws1.discourse-cdn.com/github/original/2X/4/4f48ed0636b9521a7e5f4dd857777bb9b031581a.png 1.5x, https://aws1.discourse-cdn.com/github/original/2X/4/4f48ed0636b9521a7e5f4dd857777bb9b031581a.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/4/4f48ed0636b9521a7e5f4dd857777bb9b031581a_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">Screen Shot 2020-04-17 at 12.32.06 AM.png</span><span class="informations">528×636 34.3 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>
<p>when my workflow runs i have everything setup as you said but though it takes forever to run test</p>
<p>on local it works fine but on github it takes forever</p>

**Comment by meganindya** [comment]
<p>For anybody still reading this.<br>
I faced the same issue, found that I had <code>--watch</code> flag added to <code>jest</code> in the <code>test</code> script.<br>
The test won’t end in that case. Removed and everything is fine.</p>

**Reply by jewel109** [reply]
thanks after lot of search it worked for me.

**Comment by nitin-jci** [comment]
<p>[If you are still reading this article]<br>
I have faced the same issue with angular “ng test” and followed below steps:</p>
<ol>
<li>set “singleRun: true” in karma.conf.js</li>
<li>update the ng test in package.json to “ng test --browsers=ChromeHeadless”</li>
<li>Changed the action workflow to [apt statement is for ubuntu, you can use the equivalent install statement for other platforms as well ]</li>
</ol>
<pre><code class="lang-auto">          - run: sudo apt install chromium-browser

          - name: Install dependencies
            run: npm ci

          - name: NPM Test
            run: |
              export CHROME_BIN=/usr/bin/chromium-browser
              CI=true npm run test-headless
            env:
              CI: true
</code></pre>
<p>Problem was related to the missing chromium package on ubuntu runner. After installing the chromium and exporting the environment variable, test has started running.</p>

**Comment by 22shubham22** [comment]
<p>Run sudo apt install chromium-browser</p>
<p>WARNING: apt does not have a stable CLI interface. Use with caution in scripts.</p>
<p>Reading package lists…</p>
<p>Building dependency tree…</p>
<p>Reading state information…</p>
<p>E: Unable to locate package chromium-browser</p>
<p>Error: Process completed with exit code 100.</p>
<p>cannot find chromium could you please help ?</p>

**Comment by rupinder02** [comment]
'--exit' option works for me

**Comment by Kve-sudo** [comment]
I have no fucking idea how the fuck to install pycharm on my Chromebook
I've been trying to install it for five days and I've already been driven
crazy by so many beggar commands I've entered. The furthest I've gotten it
says this page or site is available in the Linux terminal. and that pure
fucking letters and doodles.  This mother is harder to penetrate than a
Swiss bank vault.

**Comment by Yogi679** [comment]
This one worked for me. 

**Source**: https://stackoverflow.com/questions/75061237/npm-test-hangs-after-angular-unit-tests-pass-in-github-actions

"scripts": {
    "ng": "ng",
    "start": "ng serve",
    "build": "ng build",
    "watch": "ng build --watch --configuration development",
    "test": "**ng test --no-watch --source-map=false --browsers=Chrome**"
  },

**Comment by teamtunell313** [comment]
I was running `npm test` in a gitlab ci job for a node express api when I encountered this problem.
Removing `--detectOpenHandles` from my npm test command solved it for me. The other suggestions here weren't necessary without that.
